src/parsing-1.html

```diff
@@ -1,54 +1,24 @@
-         numeric
-         character references, for example &amp;#x222B; rather than &amp;int;</p>
-      
-      
-      <p>If a MathML fragment is parsed without a DTD, in other words as a well-formed XML
-         fragment, it is the responsibility of the processing application to treat the white
-         space
-         characters occurring outside of token elements as not significant.</p>
-      
-      
-      <p>However, in many circumstances, especially while producing or editing MathML, it is
-         useful to use a language definition to constrain the editing process or to check the
-         correctness of generated files. The following section, <a href="#parsing_usingrnc"></a>,
-         discusses the RelaxNG Schema for Mathml4 [[RELAX-NG]], which forms a normative
-         part of the specification. Following that, <a href="#parsing_usingxsd"></a>, and <a href="#parsing_usingdtd"></a> discuss alternative languages definition using the document type
-         definitions (DTD) and the W3C XML schema language, [[XMLSchemas]], both of
-         which are derived from the normative RelaxNG schema automatically.  One should note
-         that
-         the schema definitions of the language is currently stricter than the DTD version.
-         That
-         is, a schema validating processor will declare invalid documents that are declared
-         valid
-         by a (DTD) validating XML parser. This is partly due to the fact that the XML schema
-         language may express additional constraints not expressable in the DTD, and partly
-         due to
-         the fact that for reasons of compatibility with earlier releases, the DTD is intentionally
-         forgiving in some places and does not enforce constraints that are specified in the
-         text
-         of this specification.</p>
-      </section>
+  <h3 id="parsing_wellformed">Validating MathML</h3>
+      
+ <p>Presented here are Relax NG schema for MathML.
+ A Relax NG Schema is most naturally associated with the XML serialization
+ of MathML, but may also form the basis of  validation in other
+ formats. (Some HTML validators are based on Relax NG for
+ example).</p>
+ <p>Even when using the XML serialization, some normalization of
+ the input may be required before applying these schema, Notably,
+ following HTML, [[MathML-Core]] allows attributes such as
+ <code>onclick</code> to be specified in any case, eg
+ <code>OnClick="..."</code>
+ It is not practically feasible to specify that attribute names are
+ case insensitive here so only the lowercase names are allowed.
+ Similarly any attribute with name starting with the prefix
+ <code>data-</code> should be considered valid. The schema here only
+ allows a fixed attribute, <code>data-extra</code>,so input should be
+ normalized to remove data attributes before validating, or the schema
+ should be extended to support th attributes used in a particular
+ application.</p>
```

Comment:
I think this should be rewritten in imperative style. Maybe:
Before validating MathML with the RelaxNG schema, the following transformation have to be applied. 
1.)
2.)...
An example program is available from Appendix x

Reply:
@physikerwelt I can try to make this clearer but I was trying to avoid giving a specific imperative normalisation.

The only thing that you could specify that "always worked" would be to delete data-foo attributes before validation, however my main use of Relax NG is to drive emacs context sesitive editing, so working on the live real document. So pre-processing by removing attributes really isn't an option.  In practice if I  was (as in fact I am) working on a document format making use of data-long and data-short attributes, I would not pre-process to normalise the input, I'd use a modified local schema that allowed (just) these two data- attributes.  The case with onclick= and friends is similiar: I could specify a normalisation to lowercase so they validated but that isn't useful in editors doing live validation of the edited file.  Again I'll leave this comment open and see if I (or anyone:-) can suggest better wording.

Incidentally I think this issue means that we should not make the schema normative as it was in mathml3, I don't see how a schema that you have to use "with care" in this way, either extending the schema or modifying the document can really be normative.

Reply:
You can still delete it, if you copy the if you copy the stream and do the bookkeeping for the positions of the copied stream without MathML elements and the rest. In the same way, one could do validation for the CSS part of the style attribute, without to confuse CSS validators with MathML. We used this method in plagiarism detection extensively. However, I think it is generalizable as discussed in https://www.gipp.com/wp-content/papercite-data/pdf/beck2021.pdf 

However, I think the primary focus of the document is to clearly define what is valid MathML. A discussion of how it can be used might be better placed at a website, a forum or another resource that can be changed without the effort of creating a new revision of the standard.

I am uncertain about the question if the schema should be normative. Overall, the "code is law" paradigm has some attractively, but maybe it creates more problem than it resolves.

Reply:
@physikerwelt  yes I agree with what you write here there are certainly ways to use this in most pipelines but the spec isn't really the place to describe the processing. In practice though I am not going to go into the editor code and change its processing pipeline to remove data-foo attributes while bookeeping the visible edit buffer. That would be a lot of work and a maintenance nightmare, compared to adding 2 lines to a local copy of the schema to add the specific data-foo atributes that I want. If I am not prepared to use the schema as written and would use a locally modified version then I'm wary of specifying that that is what people  should do.

Previous versions of the spec could live in an XML world and just assert the schema was normative and any non valid document was by definition not mathml and out of scope for the specification. But in an HTML world that isn't really an option. The html parser (and mathml-core) being valid or not is just a state that is not visible to most users and the procssing is defined for all possible input.

Most MathML will (probably) live in HTML+MathML documents and there is no longer (as far as I know) an official HTML Schema, just various schemas that approximate the HTML parsing algorithm to various extents, so (a) most mathml in the wild may end up being invalid, and (b) any validation tools will need to use some custom code to validate html+mahml that may or may not make use of the schema presented here. But it's still not clear to me how much one can or should say in the spec.  I have no particular attachment to the current wording though, I'll try again later to re-word and address some of your points.

Reply:
Mh, I get the feeling that we are trying to solve the same problem that for example SVG has as well. In the abstract, the SVG spec says
> SVG content is stylable, scalable to different display resolutions, and can be viewed stand-alone, mixed with HTML content, or embedded using XML namespaces within other XML languages.

However, I could not find a RelaxNG schema there. From a first glance, I get the impression that they use IDL instead.

Reply:
yes mathml-core references (a rather minimal) webidl as well https://w3c.github.io/mathml-core/#dom-and-javascript but IDL is explicitly about software interfaces and explicitly not about the surface syntax. So that's part of the general html move away from grammars and xml view of the world.  If the html parser generates a dom with normalised case of attribute (and element) names then consuming software doesn't need to worry about that but here (if we use a validation grammar at all) we need to either make the grammar case insensitive in places (which means using a non standard schema language) or use words to authorise a certain amount of pre-validation normalisation and/or schema extension to validate documents that are making use of that freedom.

Reply:
@davidcarlisle are their major drawbacks in making relax non-normative and replace it with IDL?

Reply:
I don't think IDL is feasable at all. It's essentially a programming API so for MathML-core in the specific context of a web browser that makes sense but we don't have a DOM API for mathml full at all, we tried to specify one (and had IDL for it) in MathML2 see https://www.w3.org/TR/MathML2/appendixd.html#dom.interfaces  But  MathML Full isn't a purely web platform language and not all (or even most) implementations are DOM based so it was dropped. Adding one now is certainly put of scope for this iteration of the spec. @physikerwelt

Reply:
I'm guessing, but I think the W3C Validator, and/or epubcheck need a schema to validate, and indeed use one of the RelaxNG schemas when it comes to SVG. Random links: nu html checker [here](https://github.com/validator/validator/search?q=svg), epubcheck  [here](https://github.com/w3c/epubcheck/issues/779#issuecomment-453713182).

They seem to validate against SVG 1.1. 
The most recent schema version I could find was the RNG for SVG 1.2 [here](https://www.w3.org/Graphics/SVG/1.2/rng/). 

LaTeXML's schema is also maintained in RelaxNG, for what that's worth.

Reply:
Okay. The MathML 2 example does not look bad. Anyhow, I see that it cannot be translated from one to the other fully automatic. I wonder if we could specify a validation rule for the intent attribute (with either language).

Reply:
I was wondering about intent, relax (and xsd) schema gives you a regex pattern so you could do a light validation of intent but not actually check the full grammar.